### ISSUE TITLE
kubeadm with crio: pods with host namespace get their IP set to 127.0.0.1

### ISSUE THREAD

**Issue by mcastelino** [kubernetes-sigs/cri-o#1895]
**Description**

When setting up kubernetes with crio using kubeadm the host namespace pods get incorrect IP address. They get the IP address 127.0.0.1 instead of the IP address of the node. I happen to be using Calico, but may not be related.

**Steps to reproduce the issue:**

1. sudo -E kubeadm init --pod-network-cidr=192.168.0.0/16 --cri-socket="unix:///var/run/crio/crio.sock"
2. kubectl apply -f rbac-kdd.yaml
3. kubectl apply -f calico.yaml
4. Install other kubernetes addons (optional)

**Results:**

```
kubectl get po --all-namespaces  -o wide


ingress-nginx   nginx-ingress-controller-76f97b74b-kdlz5   1/1     Running   0          69s   192.168.0.209   crio-kata-vm   <none>
kube-system     calico-node-dr2qt                          2/2     Running   0          72s   127.0.0.1       crio-kata-vm   <none>
kube-system     coredns-576cbf47c7-nhtc7                   1/1     Running   0          72s   192.168.0.206   crio-kata-vm   <none>
kube-system     coredns-576cbf47c7-qcjdg                   1/1     Running   0          72s   192.168.0.205   crio-kata-vm   <none>
kube-system     elasticsearch-logging-0                    1/1     Running   0          72s   192.168.0.210   crio-kata-vm   <none>
kube-system     elasticsearch-logging-1                    1/1     Running   0          49s   192.168.0.214   crio-kata-vm   <none>
kube-system     etcd-crio-kata-vm                          1/1     Running   0          19s   127.0.0.1       crio-kata-vm   <none>
kube-system     fluentd-es-v2.2.1-wq2qp                    1/1     Running   0          63s   192.168.0.203   crio-kata-vm   <none>
kube-system     kibana-logging-f6fc77549-d59j6             1/1     Running   0          72s   192.168.0.212   crio-kata-vm   <none>
kube-system     kube-apiserver-crio-kata-vm                1/1     Running   0          39s   127.0.0.1       crio-kata-vm   <none>
kube-system     kube-controller-manager-crio-kata-vm       1/1     Running   0          21s   127.0.0.1       crio-kata-vm   <none>
kube-system     kube-proxy-kddwd                           1/1     Running   0          72s   127.0.0.1       crio-kata-vm   <none>
kube-system     kube-scheduler-crio-kata-vm                1/1     Running   0          11s   127.0.0.1       crio-kata-vm   <none>
kube-system     kubernetes-dashboard-77fd78f978-8jhsb      1/1     Running   0          72s   192.168.0.204   crio-kata-vm   <none>
kube-system     metrics-server-dd8c5db6-6fxz7              1/1     Running   0          72s   192.168.0.211   crio-kata-vm   <none>
monitoring      alertmanager-main-0                        2/2     Running   0          56s   192.168.0.213   crio-kata-vm   <none>
monitoring      alertmanager-main-1                        2/2     Running   0          46s   192.168.0.216   crio-kata-vm   <none>
monitoring      alertmanager-main-2                        2/2     Running   0          41s   192.168.0.217   crio-kata-vm   <none>
monitoring      grafana-5c5d49b4c4-kxclz                   1/1     Running   0          72s   192.168.0.207   crio-kata-vm   <none>
monitoring      kube-state-metrics-668bd94f74-g5gxw        4/4     Running   0          35s   192.168.0.219   crio-kata-vm   <none>
monitoring      node-exporter-7cvq7                        2/2     Running   0          63s   127.0.0.1       crio-kata-vm   <none>
monitoring      prometheus-k8s-0                           3/3     Running   1          49s   192.168.0.215   crio-kata-vm   <none>
monitoring      prometheus-k8s-1                           3/3     Running   1          38s   192.168.0.218   crio-kata-vm   <none>
monitoring      prometheus-operator-88fcf6d95-mzgc9        1/1     Running   0          72s   192.168.0.202   crio-kata-vm   <none>
```

**Describe the results you expected:**

The pods with 127.0.0.1 should have the IP address of the host

**Output of `crio --version`:**

```
crio version 1.12.0
commit: "774a29ecf6855f2dff266dc2aa2fe81d7d964465"
```

**Additional environment details:**
This issue occurs on Ubuntu 18.04 on GCP
```
$ uname -r
4.15.0-1023-gcp

$ cat /etc/lsb-release 
DISTRIB_ID=Ubuntu
DISTRIB_RELEASE=18.04
DISTRIB_CODENAME=bionic
DISTRIB_DESCRIPTION="Ubuntu 18.04.1 LTS"

$ kubelet --version
Kubernetes v1.12.2

$ sudo crio --version
crio version 1.12.0
commit: "774a29ecf6855f2dff266dc2aa2fe81d7d964465"

```

**Comment by mrunalp** [comment]
@mcastelino Do you have `manage_network_ns_lifecycle` set to true in your crio.conf?

**Comment by kelvk** [comment]
I am seeing this on my setup too. After restarting crio and kubelet, pods will be assigned with correct IP. However, a subsequent pod recreation will still cause the pod to be assigned 127.0.0.1 again.

Setting `manage_network_ns_lifecycle` to both `true` and `false` does not seem to have any effect on this behavior.

```
NAMESPACE     NAME                                       READY   STATUS    RESTARTS   AGE    IP              NODE        NOMINATED NODE
default       test-deployment-69c56d96dc-w8jhp           1/1     Running   0          39m    127.0.0.1       c2-node-1   <none>
kube-system   calico-kube-controllers-677b6c47d8-qq9fp   1/1     Running   0          161m   192.168.25.23   c2-node-1   <none>
kube-system   calico-node-cgbfm                          2/2     Running   0          100m   192.168.25.22   c2-node-0   <none>
kube-system   calico-node-j85zr                          2/2     Running   0          100m   192.168.25.23   c2-node-1   <none>
kube-system   coredns-69c9b78c7d-dwq4m                   1/1     Running   0          159m   10.242.147.67   c2-node-1   <none>
kube-system   etcd-c2-node-0                             1/1     Running   2          161m   192.168.25.22   c2-node-0   <none>
kube-system   kube-apiserver-c2-node-0                   1/1     Running   0          161m   192.168.25.22   c2-node-0   <none>
kube-system   kube-controller-manager-c2-node-0          1/1     Running   0          161m   192.168.25.22   c2-node-0   <none>
kube-system   kube-proxy-48qkt                           1/1     Running   0          3m31s   127.0.0.1       c2-node-1   <none>
kube-system   kube-proxy-c75jd                           1/1     Running   0          3m14s   192.168.25.22   c2-node-0   <none>
kube-system   kube-scheduler-c2-node-0                   1/1     Running   0          161m   192.168.25.22   c2-node-0   <none>
```

```
# uname -r
4.4.0-133-generic
# lsb_release -a
No LSB modules are available.
Distributor ID: Ubuntu
Description:    Ubuntu 16.04.5 LTS
Release:        16.04
Codename:       xenial
# kubelet --version
Kubernetes v1.12.2
# crio --version
crio version 1.12.1-dev
commit: ""
```

The Calico CNI configuration `/etc/cni/net.d/10-calico.conflist`:
```
{
  "name": "k8s-pod-network",
  "cniVersion": "0.3.0",
  "plugins": [
    {
      "type": "calico",
      "log_level": "debug",
      "etcd_endpoints": "https://192.168.25.22:2379",
      "etcd_key_file": "/etc/cni/net.d/calico-tls/etcd-key",
      "etcd_cert_file": "/etc/cni/net.d/calico-tls/etcd-cert",
      "etcd_ca_cert_file": "/etc/cni/net.d/calico-tls/etcd-ca",
      "mtu": 1440,
      "ipam": {
          "type": "calico-ipam"
      },
      "policy": {
          "type": "k8s"
      },
      "kubernetes": {
          "kubeconfig": "/etc/cni/net.d/calico-kubeconfig"
      }
    },
    {
      "type": "portmap",
      "snat": true,
      "capabilities": {"portMappings": true}
    }
  ]
}
```

Inspection of a pod with correct IP assignment after restarting crio/kubelet.
```
# crictl inspectp 89d27b2d62b24
{
  "status": {
    "id": "89d27b2d62b24ad30f3ffccabdb93037bc592b43c2989c7a07f37f04f9eb9d24",
    "metadata": {
      "attempt": 0,
      "name": "kube-proxy-c75jd",
      "namespace": "kube-system",
      "uid": "83b11bd2-ec56-11e8-8172-5254000d59e9"
    },
    "state": "SANDBOX_READY",
    "createdAt": "2018-11-19T23:54:56.0334187Z",
    "network": {
      "ip": ""
    },
    "linux": {
      "namespaces": {
        "options": {
          "ipc": "POD",
          "network": "NODE",
          "pid": "CONTAINER"
        }
      }
    },
    "labels": {
      "controller-revision-hash": "666f448674",
      "io.kubernetes.container.name": "POD",
      "io.kubernetes.pod.name": "kube-proxy-c75jd",
      "io.kubernetes.pod.namespace": "kube-system",
      "io.kubernetes.pod.uid": "83b11bd2-ec56-11e8-8172-5254000d59e9",
      "k8s-app": "kube-proxy",
      "pod-template-generation": "1"
    },
    "annotations": {
      "kubernetes.io/config.seen": "2018-11-19T23:54:55.56869521Z",
      "kubernetes.io/config.source": "api"
    }
  },
  "version": "1.12.1-dev"
}
```

Inspection of a pod with incorrect IP assignment without restarting crio/kubelet.
```
# crictl inspectp ceef7ffddf2f3
{
  "status": {
    "id": "ceef7ffddf2f38ddbc0d35916c45ca5a9655abf169ce1fed9ad5ba60cd0af27a",
    "metadata": {
      "attempt": 0,
      "name": "kube-proxy-48qkt",
      "namespace": "kube-system",
      "uid": "793ec95a-ec56-11e8-8172-5254000d59e9"
    },
    "state": "SANDBOX_READY",
    "createdAt": "2018-11-19T23:54:38.520716791Z",
    "network": {
      "ip": "127.0.0.1"
    },
    "linux": {
      "namespaces": {
        "options": {
          "ipc": "POD",
          "network": "NODE",
          "pid": "CONTAINER"
        }
      }
    },
    "labels": {
      "controller-revision-hash": "666f448674",
      "io.kubernetes.container.name": "POD",
      "io.kubernetes.pod.name": "kube-proxy-48qkt",
      "io.kubernetes.pod.namespace": "kube-system",
      "io.kubernetes.pod.uid": "793ec95a-ec56-11e8-8172-5254000d59e9",
      "k8s-app": "kube-proxy",
      "pod-template-generation": "1"
    },
    "annotations": {
      "kubernetes.io/config.seen": "2018-11-19T23:54:38.044944936Z",
      "kubernetes.io/config.source": "api"
    }
  },
  "version": "1.12.1-dev"
}
```

**Comment by mrunalp** [comment]
@kelvk For the case where you are seeing IP set to 127.0.0.1, can you check if the pod is getting its own network namespace for some reason?

**Comment by mrunalp** [comment]
We found the issue and a fix is on the way. Thanks!

**Comment by mrunalp** [comment]
This was fixed by #1924
